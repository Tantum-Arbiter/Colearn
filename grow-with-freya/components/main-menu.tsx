import React, { useState, useEffect, useCallback, useRef, useMemo } from 'react';
import { View, Pressable, StyleSheet } from 'react-native';
import { LinearGradient } from 'expo-linear-gradient';
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  cancelAnimation,
} from 'react-native-reanimated';
import { useSafeAreaInsets } from 'react-native-safe-area-context';

import { ThemedText } from './themed-text';
import { useAppStore } from '@/store/app-store';
import { DefaultPage } from './default-page';
import { ErrorBoundary } from './error-boundary';
import {
  useMemoryMonitor,
  performanceLogger,
  useSafeState
} from './main-menu/performance-utils';

import {
  MenuIcon,
  Cloud1,
  Cloud2,

  BearImage,
  ANIMATION_TIMINGS,
  LAYOUT,
  VISUAL_EFFECTS,
  DEFAULT_MENU_ITEMS,
  SCREEN_HEIGHT,
  ASSET_DIMENSIONS,
  createCloudAnimation,

  generateStarPositions,
  swapArrayItems,
  findMenuItemIndex,
  debounce,
  mainMenuStyles,
} from './main-menu/index';
import type { MenuItemData, IconStatus } from './main-menu/index';

interface MainMenuProps {
  onNavigate: (destination: string) => void;
}

function MainMenuComponent({ onNavigate }: MainMenuProps) {
  const insets = useSafeAreaInsets();

  // Performance monitoring
  useMemoryMonitor('MainMenu');

  // Get persistent animation state from store
  const { backgroundAnimationState, updateBackgroundAnimationState } = useAppStore();

  // Safe state management to prevent updates on unmounted components
  const [selectedIcon, setSelectedIcon] = useSafeState('stories-icon');
  const [menuItems, setMenuItems] = useSafeState(DEFAULT_MENU_ITEMS);
  const [triggerSelectionAnimation, setTriggerSelectionAnimation] = useSafeState(false);

  // Animation values with cleanup - initialize from persistent state
  const starRotation = useSharedValue(0);
  const cloudFloat1 = useSharedValue(backgroundAnimationState.cloudFloat1);
  const cloudFloat2 = useSharedValue(backgroundAnimationState.cloudFloat2);


  // Animation cancellation flag
  const animationsCancelled = useRef(false);

  // Safe menu state management
  const [menuOrder, setMenuOrder] = useSafeState<MenuItemData[]>(menuItems);
  const [currentPage, setCurrentPage] = useSafeState<MenuItemData | null>(null);
  const [lastSwapTime, setLastSwapTime] = useSafeState<number>(0);
  const [newlySelectedItem, setNewlySelectedItem] = useSafeState<string | null>(null);

  const timeoutRef = useRef<NodeJS.Timeout | null>(null);
  const animationWatchdogRef = useRef<NodeJS.Timeout | null>(null);

  // Component cleanup on unmount
  useEffect(() => {
    return () => {
      // Save current animation state before unmounting
      updateBackgroundAnimationState({
        cloudFloat1: cloudFloat1.value,
        cloudFloat2: cloudFloat2.value,
        rocketFloat1: 1000, // Static value - rockets removed
        rocketFloat2: -200, // Static value - rockets removed
      });

      // Cancel all animations to prevent memory leaks
      animationsCancelled.current = true;
      // Removed cancelMenuAnimation - was blocking functionality

      // Clear any pending timeouts
      if (timeoutRef.current) {
        clearTimeout(timeoutRef.current);
      }
      if (animationWatchdogRef.current) {
        clearTimeout(animationWatchdogRef.current);
      }

      // Animation limiter removed - no longer needed

      // Cancel all infinite background animations (safe for test environment)
      try {
        cancelAnimation(starRotation);
        cancelAnimation(cloudFloat1);
        cancelAnimation(cloudFloat2);
      } catch (error) {
        // cancelAnimation might not be available in test environment
        console.warn('Could not cancel background animations:', error);
      }

      const isDev = typeof __DEV__ !== 'undefined' ? __DEV__ : process.env.NODE_ENV === 'development';
      if (isDev) {
        console.log('MainMenu component unmounted - saved animation state and cleaned up');
      }
    };
  }, [updateBackgroundAnimationState]); // Removed useSharedValue objects from deps to prevent infinite re-renders

  // Performance-optimized icon press handler with debouncing and error handling
  const handleIconPressInternal = useCallback((selectedItem: MenuItemData) => {
    const endTimer = performanceLogger.startTimer('icon-press');

    try {
      // Removed startMenuAnimation check - was blocking core functionality

      const centerItem = menuOrder[0];
      const currentTime = Date.now();

      // Remove debug logging in production
      const isDev = typeof __DEV__ !== 'undefined' ? __DEV__ : process.env.NODE_ENV === 'development';
      if (isDev) {
        console.log('Icon press:', selectedItem.label);
      }

      if (selectedItem.destination === centerItem.destination) {
        // For stories, use external navigation to prevent state conflicts
        if (selectedItem.destination === 'stories') {
          onNavigate(selectedItem.destination);
        } else {
          setCurrentPage(selectedItem);
        }
      } else {
        if (currentTime - lastSwapTime < 100) {
          return; // Ignore rapid taps
        }

        const clickedIndex = menuOrder.findIndex(item => item.destination === selectedItem.destination);

        if (clickedIndex > 0) {
          setLastSwapTime(currentTime);

          const newOrder = [...menuOrder];
          [newOrder[0], newOrder[clickedIndex]] = [newOrder[clickedIndex], newOrder[0]];

          setMenuOrder(newOrder);
          setNewlySelectedItem(selectedItem.destination);

          if (timeoutRef.current) {
            clearTimeout(timeoutRef.current);
          }
          timeoutRef.current = setTimeout(() => {
            setNewlySelectedItem(null);
            // Removed endMenuAnimation - was blocking functionality
          }, 1000);
        }
      }
    } catch (error) {
      console.error('Error in handleIconPress:', error);
    } finally {
      endTimer();
    }
  }, [menuOrder, lastSwapTime, onNavigate]);

  const handleIconPress = handleIconPressInternal;

  const handleBackToMenu = () => {
    setCurrentPage(null);
  };


  useEffect(() => {
    // Small delay to ensure component is fully mounted before starting animations
    const startAnimations = () => {
      if (!animationsCancelled.current) {
        const isDev = typeof __DEV__ !== 'undefined' ? __DEV__ : process.env.NODE_ENV === 'development';

        if (isDev) {
          console.log('MainMenu: Starting animations resume');
          console.log('Cloud1 current position:', cloudFloat1.value);
          console.log('Cloud2 current position:', cloudFloat2.value);
        }

        const cloud1Pos = cloudFloat1.value;
        const cloud2Pos = cloudFloat2.value;

        // Check if positions are valid for resuming
        const cloud1CanResume = isFinite(cloud1Pos) && !isNaN(cloud1Pos) && cloud1Pos > -500 && cloud1Pos < 2000;
        const cloud2CanResume = isFinite(cloud2Pos) && !isNaN(cloud2Pos) && cloud2Pos > -500 && cloud2Pos < 2000;

        if (isDev) {
          console.log('Resume check - Cloud1:', cloud1CanResume, 'Cloud2:', cloud2CanResume);
        }

        cloudFloat1.value = createCloudAnimation(cloudFloat1, 0, -200, cloud1CanResume);
        cloudFloat2.value = createCloudAnimation(cloudFloat2, ANIMATION_TIMINGS.CLOUD_STAGGER_DELAY, -400, cloud2CanResume);

        // Rockets removed entirely

        // Set up animation watchdog to detect stuck animations
        if (animationWatchdogRef.current) {
          clearTimeout(animationWatchdogRef.current);
        }

        animationWatchdogRef.current = setTimeout(() => {
          const cloud1Pos = cloudFloat1.value;
          const cloud2Pos = cloudFloat2.value;

          if (isDev) {
            console.log('Animation watchdog check - Cloud1:', cloud1Pos, 'Cloud2:', cloud2Pos);
          }

          // If either cloud hasn't moved significantly, restart animations
          const cloud1Stuck = !isFinite(cloud1Pos) || isNaN(cloud1Pos);
          const cloud2Stuck = !isFinite(cloud2Pos) || isNaN(cloud2Pos);

          if (cloud1Stuck || cloud2Stuck) {
            if (isDev) {
              console.log('Detected stuck animations, restarting...');
            }
            // Force restart with fresh positions
            cloudFloat1.value = createCloudAnimation(cloudFloat1, 0, -200, false);
            cloudFloat2.value = createCloudAnimation(cloudFloat2, ANIMATION_TIMINGS.CLOUD_STAGGER_DELAY, -400, false);
          }
        }, 10000); // Check after 10 seconds
      }
    };

    // Start animations immediately, but allow for component mounting
    const timeoutId = setTimeout(startAnimations, 50);

    return () => {
      clearTimeout(timeoutId);
      if (animationWatchdogRef.current) {
        clearTimeout(animationWatchdogRef.current);
      }
    };
  }, []); // Run once on mount



  const starAnimatedStyle = useAnimatedStyle(() => ({
    transform: [{ rotate: `${starRotation.value}deg` }],
  }));

  const cloudAnimatedStyle1 = useAnimatedStyle(() => ({
    transform: [{ translateX: cloudFloat1.value }],
  }));

  const cloudAnimatedStyle2 = useAnimatedStyle(() => ({
    transform: [{ translateX: cloudFloat2.value }],
  }));

  // Rocket animations removed entirely

  // Generate star positions only once and keep them consistent
  // IMPORTANT: This must be called before any conditional returns to follow Rules of Hooks
  const stars = useMemo(() => generateStarPositions(), []);

  if (currentPage) {
    return (
      <DefaultPage
        icon={currentPage.icon}
        title={currentPage.label}
        onBack={handleBackToMenu}
      />
    );
  }

  return (
    <LinearGradient
      colors={VISUAL_EFFECTS.GRADIENT_COLORS}
      style={mainMenuStyles.container}
      testID="main-menu-container"
    >

      <View style={mainMenuStyles.backgroundGradient}>

        {stars.map((star) => (
          <Animated.View
            key={`star-${star.id}`}
            style={[
              mainMenuStyles.star,
              starAnimatedStyle,
              {
                left: star.left,
                top: star.top,
                opacity: star.opacity,
              }
            ]}
          />
        ))}

        <Animated.View style={[mainMenuStyles.cloudContainer, cloudAnimatedStyle1, {
          top: SCREEN_HEIGHT * LAYOUT.CLOUD_TOP_POSITION_1,
          zIndex: LAYOUT.Z_INDEX.CLOUDS_BEHIND
        }]}>
          <Cloud1 width={ASSET_DIMENSIONS.cloud1.width} height={ASSET_DIMENSIONS.cloud1.height} />
        </Animated.View>
        <Animated.View style={[mainMenuStyles.cloudContainerFront, cloudAnimatedStyle2, {
          top: SCREEN_HEIGHT * LAYOUT.CLOUD_TOP_POSITION_2,
          zIndex: LAYOUT.Z_INDEX.CLOUDS_FRONT
        }]}>
          <Cloud2 width={ASSET_DIMENSIONS.cloud2.width} height={ASSET_DIMENSIONS.cloud2.height} />
        </Animated.View>


        {/* Rockets removed entirely */}
      </View>

      <View style={mainMenuStyles.bearContainer} pointerEvents="none">
        <BearImage />
      </View>


      <View style={[legacyStyles.topButtons, { paddingTop: insets.top + 20 }]}>
        <View style={{ flex: 1 }} />

        <Pressable
          style={legacyStyles.settingsButton}
          onPress={() => setCurrentPage({ icon: 'gear', label: 'Settings', destination: 'settings' })}
        >
          <ThemedText style={mainMenuStyles.settingsEmoji}>⚙️</ThemedText>
        </Pressable>
      </View>


      <View style={legacyStyles.menuContainer}>
        <View style={mainMenuStyles.centerIcon}>
          <MenuIcon
            key={`center-${menuOrder[0].destination}`}
            icon={menuOrder[0].icon}
            label={menuOrder[0].label}
            status="animated_interactive"
            onPress={() => handleIconPress(menuOrder[0])}
            isLarge={true}
            triggerSelectionAnimation={newlySelectedItem === menuOrder[0].destination}
            testID="menu-icon-0"
          />
        </View>

        <View style={mainMenuStyles.menuContainer}>

          <View style={mainMenuStyles.topRow}>
            <MenuIcon
              key={`top-left-${menuOrder[1].destination}`}
              icon={menuOrder[1].icon}
              label={menuOrder[1].label}
              status="inactive"
              onPress={() => handleIconPress(menuOrder[1])}
              testID="menu-icon-1"
            />
            <MenuIcon
              key={`top-right-${menuOrder[2].destination}`}
              icon={menuOrder[2].icon}
              label={menuOrder[2].label}
              status="inactive"
              onPress={() => handleIconPress(menuOrder[2])}
              testID="menu-icon-2"
            />
          </View>

          {/* Bottom row */}
          <View style={mainMenuStyles.bottomRow}>
            <MenuIcon
              key={`bottom-left-${menuOrder[3].destination}`}
              icon={menuOrder[3].icon}
              label={menuOrder[3].label}
              status="inactive"
              onPress={() => handleIconPress(menuOrder[3])}
              testID="menu-icon-3"
            />
            <MenuIcon
              key={`bottom-right-${menuOrder[4].destination}`}
              icon={menuOrder[4].icon}
              label={menuOrder[4].label}
              status="inactive"
              onPress={() => handleIconPress(menuOrder[4])}
              testID="menu-icon-4"
            />
          </View>
        </View>
      </View>
    </LinearGradient>
  );
}

const legacyStyles = StyleSheet.create({
  topButtons: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingHorizontal: 20,
    paddingBottom: 20,
  },
  settingsButton: {
    backgroundColor: 'rgba(255, 255, 255, 0.95)',
    padding: 14,
    borderRadius: 22,
    borderWidth: 2,
    borderColor: 'rgba(46, 139, 139, 0.3)',
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 3 },
    shadowOpacity: 0.2,
    shadowRadius: 6,
    elevation: 6,
  },
  menuContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    paddingHorizontal: 20,
    paddingTop: 20,
  },
});

// Wrap with error boundary for crash protection
const MainMenuWithErrorBoundary = React.memo(function MainMenuWithErrorBoundary(props: MainMenuProps) {
  return (
    <ErrorBoundary
      onError={(error, errorInfo) => {
        console.error('MainMenu crashed:', error, errorInfo);
        // In production, send to crash reporting service
      }}
    >
      <MainMenuComponent {...props} />
    </ErrorBoundary>
  );
});

export const MainMenu = MainMenuWithErrorBoundary;
