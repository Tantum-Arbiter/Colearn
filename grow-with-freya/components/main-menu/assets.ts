export const PNG_ASSETS = {
  bear: require('../../assets/images/ui-elements/bear-bottom-screen.png'),
  freyaRocket: require('../../assets/images/ui-elements/freya-rocket.svg'),
  freyaRocketRight: require('../../assets/images/ui-elements/freya-rocket-right.svg'),
} as const;

export const SVG_PATHS = {
  cloud: 'ui-elements/background-cloud-1.svg',
  stories: 'menu-icons/stories-icon.svg',
  sensory: 'menu-icons/sensory-icon.svg',
  emotions: 'menu-icons/emotions-icon.svg',
  bedtime: 'menu-icons/bedtime-icon.svg',
  screentime: 'menu-icons/screentime-icon.svg',
} as const;

const ICON_SVG_MAP = {
  'stories-icon': 'stories',
  'sensory-icon': 'sensory',
  'emotions-icon': 'emotions',
  'bedtime-icon': 'bedtime',
  'screentime-icon': 'screentime',
  'storybook': 'stories',
  'sparkle_hand': 'sensory',
  'smiley_face': 'emotions',
  'music_note': 'bedtime',
  'clock': 'screentime',
} as const;

export const getIconSvgType = (iconType: string): keyof typeof SVG_PATHS => {
  return ICON_SVG_MAP[iconType as keyof typeof ICON_SVG_MAP] || 'stories';
};

export const getSvgPath = (svgType: keyof typeof SVG_PATHS) => {
  return SVG_PATHS[svgType];
};

export const ASSET_DIMENSIONS = {
  cloud1: { width: 140, height: 170 },
  cloud2: { width: 160, height: 190 },
  balloon1: { width: 140, height: 170 },
  balloon2: { width: 160, height: 190 },
  rocket: { width: 80, height: 60 },
  bear: { width: 286, height: 286 },
  icon: {
    small: { width: 48, height: 48 },
    medium: { width: 58, height: 58 },
    large: { width: 70, height: 70 },
  },
} as const;

export const getSvgAsset = (svgType: string) => {
  switch (svgType) {
    case 'cloud': return SVG_PATHS.cloud;
    case 'balloon': return SVG_PATHS.cloud; // Backward compatibility - balloon maps to cloud
    case 'stories': return SVG_PATHS.stories;
    case 'sensory': return SVG_PATHS.sensory;
    case 'emotions': return SVG_PATHS.emotions;
    case 'bedtime': return SVG_PATHS.bedtime;
    case 'screentime': return SVG_PATHS.screentime;
    default: return SVG_PATHS.stories;
  }
};

export const getSvgAssetFromPath = (svgType: keyof typeof SVG_PATHS) => {
  return getSvgAsset(svgType);
};

export const getSvgComponentFromSvg = (svgType: keyof typeof SVG_PATHS) => {
  const {
    CloudSvg,
    StoriesSvg,
    SensorySvg,
    EmotionsSvg,
    BedtimeSvg,
    ScreentimeSvg,
  } = require('./svg-components');

  switch (svgType) {
    case 'cloud':
    case 'balloon': // Backward compatibility - balloon maps to cloud
      return CloudSvg;
    case 'stories':
      return StoriesSvg;
    case 'sensory':
      return SensorySvg;
    case 'emotions':
      return EmotionsSvg;
    case 'bedtime':
      return BedtimeSvg;
    case 'screentime':
      return ScreentimeSvg;
    default:
      return StoriesSvg;
  }
};


export const getSvgComponent = getSvgComponentFromSvg;
