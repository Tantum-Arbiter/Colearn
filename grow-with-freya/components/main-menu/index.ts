export { MenuIcon } from './menu-icon';
export {
  Cloud1,
  Cloud2,
  BearImage,
} from './animated-components';

export {
  CloudSvg,
  StoriesSvg,
  SensorySvg,
  EmotionsSvg,
  BedtimeSvg,
  ScreentimeSvg,
} from './svg-components';

export {
  ANIMATION_TIMINGS,
  LAYOUT,
  VISUAL_EFFECTS,
  DEFAULT_MENU_ITEMS,
  SCREEN_WIDTH,
  SCREEN_HEIGHT,
  Easing,
} from './constants';
export type { MenuItemData, IconStatus } from './constants';

export { PNG_ASSETS, getIconSvgType, getSvgPath, ASSET_DIMENSIONS } from './assets';

export {
  createCloudAnimation,
  createIconPulseAnimation,
  createGlowAnimation,
  createShimmerAnimation,
  createSelectionAnimation,
  createPressAnimation,
  createGlowBurstAnimation,
} from './animations';

export {
  generateStarPositions,
  swapArrayItems,
  findMenuItemIndex,
  isValidMenuItem,
  getResponsiveSize,
  debounce,
  throttle,
  calculateSafePadding,
  logPerformance,
} from './utils';

export { mainMenuStyles, menuIconStyles, animatedElementStyles } from './styles';
