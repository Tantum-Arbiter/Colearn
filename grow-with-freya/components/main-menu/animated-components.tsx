import React, { useEffect } from 'react';
import { Image, View } from 'react-native';
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
  withDelay,
  Easing,
} from 'react-native-reanimated';
import { ASSET_DIMENSIONS } from './assets';
import { CloudSvg, FreyaRocketSvg, FreyaRocketRightSvg } from './svg-components';
import { CachedBearImage, CachedMoonImage } from '../ui/cached-image';

export const Cloud1 = React.memo(function Cloud1({
  width = ASSET_DIMENSIONS.cloud1.width,
  height = ASSET_DIMENSIONS.cloud1.height
}: {
  width?: number;
  height?: number;
}) {
  return (
    <CloudSvg
      width={width}
      height={height}
      opacity={0.8}
    />
  );
});

export const Cloud2 = React.memo(function Cloud2({
  width = ASSET_DIMENSIONS.cloud2.width,
  height = ASSET_DIMENSIONS.cloud2.height
}: {
  width?: number;
  height?: number;
}) {
  return (
    <CloudSvg
      width={width}
      height={height}
      opacity={0.6}
    />
  );
});

export const FreyaRocket = React.memo(function FreyaRocket({
  width = ASSET_DIMENSIONS.rocket.width,
  height = ASSET_DIMENSIONS.rocket.height
}: {
  width?: number;
  height?: number;
}) {
  return (
    <FreyaRocketSvg
      width={width}
      height={height}
      opacity={0.9}
    />
  );
});

export const FreyaRocketRight = React.memo(function FreyaRocketRight({
  width = ASSET_DIMENSIONS.rocket.width,
  height = ASSET_DIMENSIONS.rocket.height
}: {
  width?: number;
  height?: number;
}) {
  return (
    <FreyaRocketRightSvg
      width={width}
      height={height}
      opacity={0.9}
    />
  );
});

export const BearImage = React.memo(function BearImage() {
  return (
    <View style={{
      width: '100%',
      height: '100%',
      justifyContent: 'flex-end',
      alignItems: 'center',
    }}>
      <CachedBearImage />
    </View>
  );
});

export const MoonImage = React.memo(function MoonImage() {
  return (
    <View style={{
      width: '100%',
      height: '100%',
      justifyContent: 'flex-start',
      alignItems: 'center',
    }}>
      <CachedMoonImage />
    </View>
  );
});
