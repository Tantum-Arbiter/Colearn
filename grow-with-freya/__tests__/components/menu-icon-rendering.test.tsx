import React from 'react';
import { render } from '@testing-library/react-native';
import { MenuIcon } from '../../components/main-menu/menu-icon';
import { getSvgComponentFromSvg } from '../../components/main-menu/assets';

// Mock the SVG components to avoid rendering issues in tests
jest.mock('../../components/main-menu/svg-components', () => ({
  StoriesSvg: ({ width, height, opacity }: any) => `MockStoriesSvg-${width}x${height}-${opacity}`,
  SensorySvg: ({ width, height, opacity }: any) => `MockSensorySvg-${width}x${height}-${opacity}`,
  EmotionsSvg: ({ width, height, opacity }: any) => `MockEmotionsSvg-${width}x${height}-${opacity}`,
  BedtimeSvg: ({ width, height, opacity }: any) => `MockBedtimeSvg-${width}x${height}-${opacity}`,
  ScreentimeSvg: ({ width, height, opacity }: any) => `MockScreentimeSvg-${width}x${height}-${opacity}`,
  CloudSvg: ({ width, height, opacity }: any) => `MockCloudSvg-${width}x${height}-${opacity}`,
  BearSvg: ({ width, height, opacity }: any) => `MockBearSvg-${width}x${height}-${opacity}`,
}));

describe('MenuIcon Icon Rendering Tests', () => {
  const mockOnPress = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('SVG Component Resolution', () => {
    it('should resolve stories icon to StoriesSvg component', () => {
      const SvgComponent = getSvgComponentFromSvg('stories');
      expect(SvgComponent).toBeDefined();
      expect(typeof SvgComponent).toBe('function');
    });

    it('should resolve sensory icon to SensorySvg component', () => {
      const SvgComponent = getSvgComponentFromSvg('sensory');
      expect(SvgComponent).toBeDefined();
      expect(typeof SvgComponent).toBe('function');
    });

    it('should resolve emotions icon to EmotionsSvg component', () => {
      const SvgComponent = getSvgComponentFromSvg('emotions');
      expect(SvgComponent).toBeDefined();
      expect(typeof SvgComponent).toBe('function');
    });

    it('should resolve bedtime icon to BedtimeSvg component', () => {
      const SvgComponent = getSvgComponentFromSvg('bedtime');
      expect(SvgComponent).toBeDefined();
      expect(typeof SvgComponent).toBe('function');
    });

    it('should resolve screentime icon to ScreentimeSvg component', () => {
      const SvgComponent = getSvgComponentFromSvg('screentime');
      expect(SvgComponent).toBeDefined();
      expect(typeof SvgComponent).toBe('function');
    });

    it('should resolve cloud icon to CloudSvg component', () => {
      const SvgComponent = getSvgComponentFromSvg('cloud');
      expect(SvgComponent).toBeDefined();
      expect(typeof SvgComponent).toBe('function');
    });

    it('should resolve balloon icon to CloudSvg component (alias)', () => {
      const SvgComponent = getSvgComponentFromSvg('balloon');
      expect(SvgComponent).toBeDefined();
      expect(typeof SvgComponent).toBe('function');
    });

    it('should fallback to StoriesSvg for unknown icon types', () => {
      const SvgComponent = getSvgComponentFromSvg('unknown' as any);
      expect(SvgComponent).toBeDefined();
      expect(typeof SvgComponent).toBe('function');
    });
  });

  describe('MenuIcon Component Rendering', () => {
    it('should render stories icon without errors', () => {
      const { toJSON } = render(
        <MenuIcon
          icon="stories-icon"
          text="Stories"
          onPress={mockOnPress}
          isSelected={false}
          animationDelay={0}
        />
      );

      expect(toJSON()).toBeTruthy();
      expect(toJSON()).toMatchSnapshot();
    });

    it('should render sensory icon without errors', () => {
      const { toJSON } = render(
        <MenuIcon
          icon="sensory-icon"
          text="Sensory"
          onPress={mockOnPress}
          isSelected={false}
          animationDelay={0}
        />
      );

      expect(toJSON()).toBeTruthy();
      expect(toJSON()).toMatchSnapshot();
    });

    it('should render emotions icon without errors', () => {
      const { toJSON } = render(
        <MenuIcon
          icon="emotions-icon"
          text="Emotions"
          onPress={mockOnPress}
          isSelected={false}
          animationDelay={0}
        />
      );

      expect(toJSON()).toBeTruthy();
      expect(toJSON()).toMatchSnapshot();
    });

    it('should render bedtime icon without errors', () => {
      const { toJSON } = render(
        <MenuIcon
          icon="bedtime-icon"
          text="Bedtime"
          onPress={mockOnPress}
          isSelected={false}
          animationDelay={0}
        />
      );

      expect(toJSON()).toBeTruthy();
      expect(toJSON()).toMatchSnapshot();
    });

    it('should render screentime icon without errors', () => {
      const { toJSON } = render(
        <MenuIcon
          icon="screentime-icon"
          text="Screen Time"
          onPress={mockOnPress}
          isSelected={false}
          animationDelay={0}
        />
      );

      expect(toJSON()).toBeTruthy();
      expect(toJSON()).toMatchSnapshot();
    });
  });

  describe('Icon Visibility and Properties', () => {
    it('should render icon with correct size when selected', () => {
      const { toJSON } = render(
        <MenuIcon
          icon="stories-icon"
          text="Stories"
          onPress={mockOnPress}
          isSelected={true}
          animationDelay={0}
        />
      );
      
      const rendered = toJSON();
      expect(rendered).toBeTruthy();
      // The icon should be present in the rendered output
      expect(JSON.stringify(rendered)).toContain('Mock');
    });

    it('should render icon with correct size when not selected', () => {
      const { toJSON } = render(
        <MenuIcon
          icon="stories-icon"
          text="Stories"
          onPress={mockOnPress}
          isSelected={false}
          animationDelay={0}
        />
      );
      
      const rendered = toJSON();
      expect(rendered).toBeTruthy();
      // The icon should be present in the rendered output
      expect(JSON.stringify(rendered)).toContain('Mock');
    });

    it('should not crash when rendering unknown icon type', () => {
      expect(() => {
        render(
          <MenuIcon
            icon="unknown-icon" as any
            text="Unknown"
            onPress={mockOnPress}
            isSelected={false}
            animationDelay={0}
          />
        );
      }).not.toThrow();
    });
  });

  describe('Icon Component Function Validation', () => {
    it('should ensure all icon components are functions', () => {
      const iconTypes = ['stories', 'sensory', 'emotions', 'bedtime', 'screentime', 'cloud', 'balloon'];
      
      iconTypes.forEach(iconType => {
        const SvgComponent = getSvgComponentFromSvg(iconType as any);
        expect(typeof SvgComponent).toBe('function');
        expect(SvgComponent.name).toBeTruthy(); // Should have a component name
      });
    });

    it('should ensure icon components can be instantiated', () => {
      const iconTypes = ['stories', 'sensory', 'emotions', 'bedtime', 'screentime'];
      
      iconTypes.forEach(iconType => {
        const SvgComponent = getSvgComponentFromSvg(iconType as any);
        expect(() => {
          // Test that the component can be created without throwing
          const element = React.createElement(SvgComponent, { width: 58, height: 58, opacity: 1 });
          expect(element).toBeTruthy();
        }).not.toThrow();
      });
    });
  });
});
