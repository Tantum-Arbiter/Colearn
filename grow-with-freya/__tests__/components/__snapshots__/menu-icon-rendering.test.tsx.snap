// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuIcon Icon Rendering Tests MenuIcon Component Rendering should render bedtime icon without errors 1`] = `
<div
  className="css-view-g5y9jx r-alignItems-1awozwy r-justifyContent-1777fci"
  dir={null}
  ref={[Function]}
>
  <button
    aria-label="undefined button"
    className="css-view-g5y9jx r-cursor-1loqt21 r-touchAction-1otgn73 r-alignItems-1awozwy r-backgroundColor-6st0mw r-borderRadius-1lmcfpr r-boxShadow-12et9rd r-height-37igu r-justifyContent-1777fci r-width-19554kt"
    dir={null}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    ref={[Function]}
    role="button"
    tabIndex={0}
    type="button"
  >
    MockBedtimeSvg-48x48-1
  </button>
</div>
`;

exports[`MenuIcon Icon Rendering Tests MenuIcon Component Rendering should render emotions icon without errors 1`] = `
<div
  className="css-view-g5y9jx r-alignItems-1awozwy r-justifyContent-1777fci"
  dir={null}
  ref={[Function]}
>
  <button
    aria-label="undefined button"
    className="css-view-g5y9jx r-cursor-1loqt21 r-touchAction-1otgn73 r-alignItems-1awozwy r-backgroundColor-6st0mw r-borderRadius-1lmcfpr r-boxShadow-12et9rd r-height-37igu r-justifyContent-1777fci r-width-19554kt"
    dir={null}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    ref={[Function]}
    role="button"
    tabIndex={0}
    type="button"
  >
    MockEmotionsSvg-48x48-1
  </button>
</div>
`;

exports[`MenuIcon Icon Rendering Tests MenuIcon Component Rendering should render screentime icon without errors 1`] = `
<div
  className="css-view-g5y9jx r-alignItems-1awozwy r-justifyContent-1777fci"
  dir={null}
  ref={[Function]}
>
  <button
    aria-label="undefined button"
    className="css-view-g5y9jx r-cursor-1loqt21 r-touchAction-1otgn73 r-alignItems-1awozwy r-backgroundColor-6st0mw r-borderRadius-1lmcfpr r-boxShadow-12et9rd r-height-37igu r-justifyContent-1777fci r-width-19554kt"
    dir={null}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    ref={[Function]}
    role="button"
    tabIndex={0}
    type="button"
  >
    MockScreentimeSvg-48x48-1
  </button>
</div>
`;

exports[`MenuIcon Icon Rendering Tests MenuIcon Component Rendering should render sensory icon without errors 1`] = `
<div
  className="css-view-g5y9jx r-alignItems-1awozwy r-justifyContent-1777fci"
  dir={null}
  ref={[Function]}
>
  <button
    aria-label="undefined button"
    className="css-view-g5y9jx r-cursor-1loqt21 r-touchAction-1otgn73 r-alignItems-1awozwy r-backgroundColor-6st0mw r-borderRadius-1lmcfpr r-boxShadow-12et9rd r-height-37igu r-justifyContent-1777fci r-width-19554kt"
    dir={null}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    ref={[Function]}
    role="button"
    tabIndex={0}
    type="button"
  >
    MockSensorySvg-48x48-1
  </button>
</div>
`;

exports[`MenuIcon Icon Rendering Tests MenuIcon Component Rendering should render stories icon without errors 1`] = `
<div
  className="css-view-g5y9jx r-alignItems-1awozwy r-justifyContent-1777fci"
  dir={null}
  ref={[Function]}
>
  <button
    aria-label="undefined button"
    className="css-view-g5y9jx r-cursor-1loqt21 r-touchAction-1otgn73 r-alignItems-1awozwy r-backgroundColor-6st0mw r-borderRadius-1lmcfpr r-boxShadow-12et9rd r-height-37igu r-justifyContent-1777fci r-width-19554kt"
    dir={null}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    ref={[Function]}
    role="button"
    tabIndex={0}
    type="button"
  >
    MockStoriesSvg-48x48-1
  </button>
</div>
`;
